src/io_ops/nonblocking_io.rs

```diff
@@ -0,0 +1,91 @@
+use super::FileDescriptorRawHandle;
+use libc::O_NONBLOCK;
+use libc::{self, c_int, c_void, size_t};
+use log::error;
+use std::io;
+use std::sync::Arc;
+use tokio::io::unix::AsyncFd;
+
+#[derive(Debug, Clone)]
+pub struct SubChannel {
+    fd: Arc<AsyncFd<FileDescriptorRawHandle>>,
+}
+impl SubChannel {
+    pub fn as_raw_fd(&self) -> &FileDescriptorRawHandle {
+        self.fd.as_ref().get_ref()
+    }
+
+    pub fn new(fd: FileDescriptorRawHandle) -> io::Result<SubChannel> {
+        let code = unsafe { libc::fcntl(fd.fd, libc::F_SETFL, O_NONBLOCK) };
+        if code == -1 {
+            error!(
+                "fcntl set file handle to O_NONBLOCK command failed with {}",
+                code
+            );
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(SubChannel {
+            fd: Arc::new(AsyncFd::new(fd)?),
+        })
+    }
+    /// Send all data in the slice of slice of bytes in a single write (can block).
+    pub async fn send(&self, buffer: &[&[u8]]) -> io::Result<()> {
+        loop {
+            let mut guard = self.fd.writable().await?;
+
+            match guard.try_io(|inner| {
+                let iovecs: Vec<_> = buffer
+                    .iter()
+                    .map(|d| libc::iovec {
+                        iov_base: d.as_ptr() as *mut c_void,
+                        iov_len: d.len() as size_t,
+                    })
+                    .collect();
+                let rc = unsafe {
+                    libc::writev(inner.get_ref().fd, iovecs.as_ptr(), iovecs.len() as c_int)
+                };
+                if rc < 0 {
+                    Err(io::Error::last_os_error())
+                } else {
+                    Ok(())
+                }
+            }) {
+                Ok(result) => return result,
+                Err(_would_block) => continue,
+            }
+        }
+    }
+
+    pub fn close(&self) {
+        self.fd.get_ref().close()
+    }
+
+    pub async fn do_receive(&self, buffer: &'_ mut [u8]) -> io::Result<Option<usize>> {
+        use std::time::Duration;
+        use tokio::time::timeout;
+        loop {
+            if let Ok(guard_result) = timeout(Duration::from_millis(1000), self.fd.readable()).await
```

Comment:
Instead of a 1sec timeout, is it possible to wait on both the read and the `terminated` future that's passed into `Channel::receive`? I think that would remove any unmount delay

Reply:
the terminated future is only satisfied by one of these inner reads firing and triggering the exit cascade. -- we actually do wait on it outside of here, the outer `select!` should get satisfied based on terminated if/when it happens.

Reply:
I see, maybe I misunderstood your previous comment, "(3) combined with (1) or (2), means that the process exits 1 second after". Is that 1sec delay caused by this timeout, or something else?

Reply:
I probably didn’t explain this super well from an IO/FD perspective if we go through the last request done:

1. User/someone calls `umount` on mount point
2. Poll on socket `readiness` fires
3. `recv` on socket gets a Statfs request(for some ABI’s the kernel always issues this after an unmount it seems)
4. `readiness` will return true on next try
5.  `recv` Returns `E_WOULDBLOCK`
6.   `readiness` blocks
7.  ..some delay in kernel...
8.   FD handles become invalid
9.   Timeout fires in code
10.   Call `recv` again — returns INVALID FD i think is the error message


The issue is around step `8` we would expect `readiness` Call to return and wake things up, but this doesn’t seem to happen with fuse, since the FD isn’t closed/marked as readable by the kernel, its just invalid somehow. The 1 second timeout/timeout in general is an arbitrary workaround for this issue. (Maybe there is some lower level thing if we used epoll directly that does fire? )

Reply:
I see, and where in that sequence do we receive the `Operation::Destroy`? Since the code does `select!` on the `terminated` signal, it seems like that would happen before the timeout fires and avoid the 1sec delay

Reply:
That i've seen/maybe i've broken something... but `Operation::Destroy` is never delivered :/ Your right, it seems like it is the right thing to happen here

Reply:
Might have changed in the mean time, but https://github.com/libfuse/libfuse/commit/b9ea32fc3bb3410c5f436c1f1f90558ff189bf23 with the original changelog entry would explain never seeing it:

```
	* Add a synchronous DESTROY message to kernel interface.  This is
	invoked from umount, when the final instance of the filesystem is
	released.  It is only sent for filesystems mounted with the
	'blkdev' option for security reasons.
```

Reply:
Oh, well that's good to know! Ok, I probably need to read through some of this again with that in mind.

So, if I understand correctly then, previously we only used `libc::read` which returns immediately when the FD becomes invalid. However, now we call `readiness` and that blocks for an unknown (and long) delay after the FD is invalidated, which is why we need a timeout now. Is that right?

Reply:
yep, thats it exactly. It seems that the kernel component of fuse must just kill off blocked read operations. But whatever way it works(or possibly interacts with tokio's asyncfd stack/mio in rust), poll calls don't get woken up the same way. If the file handle got closed then the readiness would wake properly. But something 'special' seemingly about fuse happens where where it becomes invalid. so it doesn't wake.

Reply:
Yeah i had been avoiding too low a timeout, but 1ms for just one FD seems reasonable enough to me. I've added it so the root/original FD will now have a 1ms timeout. And re-enabled the mount tests. It is still a race thats possible in the code but it should be much less likely?

(Also lots of churn since i rebased on the master changes around the API separation)